src/rust/engine/src/scheduler.rs

```diff
@@ -295,64 +297,92 @@ impl Scheduler {
       .map(NodeKey::from)
       .collect();
 
-    // Lock the display for the remainder of the execution, and grab a reference to it.
-    let mut maybe_display = match &session.display() {
-      &Some(ref d) => Some(d.lock()),
-      &None => None,
-    };
-
     // This map keeps the k most relevant jobs in assigned possitions.
     // Keys are positions in the display (display workers) and the values are the actual jobs to print.
     let mut tasks_to_display = IndexMap::new();
 
-    if let Some(ref mut display) = maybe_display {
-      display.start();
-    };
-
-    let results = loop {
-      if let Ok(res) = receiver.recv_timeout(Duration::from_millis(100)) {
-        break res;
-      } else if let Some(ref mut display) = maybe_display {
-        Scheduler::display_ongoing_tasks(&self.core.graph, &roots, display, &mut tasks_to_display);
+    let results = match session.maybe_display() {
+      Some(display) => {
+        {
+          let mut display = display.lock();
+          display.start();
+        }
+        let unique_handle = LOGGER.register_engine_display(display.clone());
+
+        let results = loop {
+          if let Ok(res) = receiver.recv_timeout(Duration::from_millis(100)) {
+            break res;
+          } else {
+            Scheduler::display_ongoing_tasks(
+              &self.core.graph,
+              &roots,
+              display,
+              &mut tasks_to_display,
+            );
+          }
+        };
+        {
+          let mut display = display.lock();
+          display.finish();
+        }
+        LOGGER.deregister_engine_display(unique_handle);
```

Comment:
Should probably swap `deregister` with `display.finish()`, so that we stop receiving log messages before we finish.